Make saved insights meta fields editable in-line

alexkim205 opened the pull request:
## Changes

Closes #6131.

Should've gotten most of the edge cases with different types of inputs but please feel free to poke around and break things.

**Demo**


https://user-images.githubusercontent.com/13460330/135812797-313325f3-d5cf-4dfe-b98f-d76689d0e558.mov



## How did you test this code?

- tests for new logics
- tests for new utils

alexkim205 committed:
refactor insight metadata into separate reusable components

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-saved-insights-metadata

alexkim205 committed:
create insightMetdataLogic

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-saved-insights-metadata

alexkim205 committed:
added editable states to fields

alexkim205 committed:
description add suffix to input

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-saved-insights-metadata

alexkim205 committed:
fix functional bugs with input values

alexkim205 committed:
remove consoles

alexkim205 commented on frontend/src/lib/components/ObjectTags/index.tsx:
```diff
@@ -48,25 +48,29 @@ export function ObjectTags({
-                )
-            })}
+            {tags
+                .filter((t) => !!t)
```
Fixes an existing bug where we were trying to render undefined tags which led to white screening.

alexkim205 committed:
testing for insightMetadataLogic

alexkim205 committed:
revert Insight

alexkim205 committed:
revert Insight

pauldambra commented on frontend/src/scenes/insights/InsightMetadata/InsightMetadata.tsx:
```diff
@@ -0,0 +1,182 @@
+    )
+}
+
+function Description({ insight, isEditable = false }: MetadataProps): JSX.Element | null {
```
I wonder if there's value in refactoring to extract an `EditableTextField` component. It would reduce duplication of the idea here but could also be lifted out as a component

alexkim205 replied:
Hmm, in my first iterations I tried to refactor this part out, but I found it difficult and even more confusing to read especially with the api differences between `Input.TextArea` and `Input`. It'd be really nice if `Input.TextArea` supported the `suffix` prop so that I don't need this hacky wrapper workaround, but until then I think it'd be better to leave this for now.

pauldambra commented on frontend/src/scenes/insights/insightLogic.tsx:
```diff
@@ -84,7 +88,14 @@ export const insightLogic = kea<insightLogicType>({
             },
-            setInsight: (insight) => insight,
+            setInsight: ({ insight, shouldMergeWithExisting }) => {
+                return shouldMergeWithExisting
```
nit... we could remove braces and `return` here

pauldambra reviewed (approved):
Looks good to me 🙌 . No reason to block this merge on the idea of extracting a component from the editable text although (if nothing else) it does save the future reader comparing the title and description to understand differences

clarkus reviewed (dismissed):
I like this change. Based on the feedback during the demo today, we might consider making renaming, description editing, tagging available outside of the edit view. Is it technically feasible to save those single-attribute changes outside the edit workflow? Great work, Alex!

alexkim205 commented:
Thanks for the feedback everyone!

> Is it technically feasible to save those single-attribute changes outside the edit workflow?

Yup it should be. The only difference would be if the user clicks "Done" outside of the edit workflow, it will permanently save the metadata for that insight. This is different from behavior in the edit workflow, where clicking "Done" will just save the state of the metadata and clicking "Save" will permanently save the final state.

alexkim205 committed:
make editable in view mode

alexkim205 commented:
Added edit functionality to view modes and cleaned up FF's for this feature.

https://user-images.githubusercontent.com/13460330/135937210-5ab46468-cffb-490d-9dad-8457b436d188.mov

alexkim205 committed:
test for update path

mariusandra commented:
Overall LGTM. Going to merge since I want to work on the insight logics a bit today, and best to have it in... and it's feature flagged.

The two UX issues I found here are:
1. Editing tags inside the "view" mode doesn't save them
2. When I reload the insight, sometimes I see the edited text, sometimes not. 
3. Sometimes I get to a state where it creates a new insights on every reload

All these are proper blockers to unflagging saved insights, but not for merging this PR.
